probe-rs/src/probe/usb_util.rs

```diff
@@ -25,18 +25,25 @@ impl InterfaceExt for Interface {
     }
 
     fn read_bulk(&self, endpoint: u8, buf: &mut [u8], timeout: Duration) -> io::Result<usize> {
-        let fut = async {
-            let comp = self.bulk_in(endpoint, RequestBuffer::new(buf.len())).await;
-            comp.status.map_err(io::Error::other)?;
-
-            let n = comp.data.len();
-            buf[..n].copy_from_slice(&comp.data);
-            Ok(n)
+        let mut queue = self.bulk_in_queue(endpoint);
+        queue.submit(RequestBuffer::new(buf.len()));
+        let Some(comp) = block_on(
+            async {
+                let comp = queue.next_complete().await;
+                Some(comp)
+            }
+            .or(async {
```

Comment:
Does this really need two separate async blocks? It's really confusing how this works (or what you intended to do) - if the first async block is always Some, the second one never runs I think.

Did you intend to use some `select`-like construct here?

Reply:
The same construction is already in use before my change. I just intended to keep it that way.

Reply:
Where? We should fix that, it looks like nonsense

Reply:
@Dirbaio can you explain the `fut.or` pattern in the other function in this file? Why are we waiting a `timeout` amount of time, if an error occurs?

Reply:
`futures_lite::FutureExt::or()` is equivalent to `select`

Reply:
Oh TIL, thanks, it looks _very_ weird.

Reply:
I agree, I don't like it either. I used that because I wanted to avoid pulling in heavier deps tho.

Reply:
At least .or saves us a match on an Either, which is nice... I guess I'd appreciate a comment for dummies like me, that this isn't an `or` on the async block's result, but on the future.

Reply:
This will soon all be easier because we go async most likely. But cancelling is actually not possible in webusb and timeouts are also not as easy. Two open points to figure out.